### ISSUE TITLE
RequiredIf with input select

### ISSUE THREAD

**Issue by nunos7** [JaroslawWaliszko/ExpressiveAnnotations#46]
Hi, 
I have a module with some properties:
```
        public bool HaveHolder { get; set; }
        [RequiredIf("HaveHolder == true")]
        public int? AffinityID { get; set; }
        public Affinity Afinitty { get; set; }

        [RequiredIf("HaveHolder == true")]
        public int? HolderNIF { get; set; }
```
In View the AffinitiyID is rendered with a input select, and the HolderNIF with a input type text (numeric). The input type text fired the validation if my checkbox HaveHolder is not check, but the input select don't...

This is part of the View for the AffinityID property:
```
       <div class="form-group">
                @Html.LabelFor(model => model.AffinityID, "AffinityID", htmlAttributes: new { @class = "control-label col-md-2" })
                <div class="col-md-10">
                    @Html.DropDownList("AffinityID", null, "-- select one option --", htmlAttributes: new { @class = "form-control" })
                    @Html.ValidationMessageFor(model => model.AffinityID, "", new { @class = "text-danger" })
                </div>
            </div>
```

**Comment by JaroslawWaliszko** [comment]
Hi,
I see no issue in your sample. It should work as expected. The source of your problem must be hidden in some another place.

Actually there is one inconsistency (but I suppose unrelated) - you're not providing any items for your drop down list - doesn't it throw an exception?

*There is no ViewData item of type 'IEnumerable\<SelectListItem\>' that has the key 'AffinityID'.*

The correct construct:
`@Html.DropDownListFor(model => model.AffinityID, Model.Feeds, "--select--", new { @class = "..." })`

**Comment by nunos7** [comment]
Hi, thanks for the quick answer.
The dropdownlist data is loaded from the controller:

```
ViewBag.AffinityID = new SelectList(db.Affinitiy, "AffinityID", "Designation", client.AffinityID);
```

I inspect the javascript code and the "this.referencesMap[name]" only has the field HolderNIF, and none of the other fields (ActivityID).

I will try to make a small sample and send to you...

Nuno Santos

**Comment by JaroslawWaliszko** [comment]
No problem. The minimal but working sample will be the easiest for me to track down your issue.

**Comment by nunos7** [comment]
Hi, 
I  made available the demo code in this link:
http://nsantos.eu/EASample.7z

Thanks,
Nuno Santos

**Comment by JaroslawWaliszko** [comment]
Hi, I've opened this but the project does not compile, some of files are missing etc. If you find some more time, try give me something which I can build and run, otherwise it will be hard for me to find spare time for that. Data can be obviously mocked.

Regards

**Comment by nunos7** [comment]
Ok I will try to share the full project in dropbox.

**Comment by JaroslawWaliszko** [comment]
In general, you should use strongly typed helpers instead of weakly typed ones. For some reason when using weakly typed helper:

```@Html.DropDownList("CarModelID", null, "-- select --", htmlAttributes: new { @class = "form-control" })```

ASP.MVC framework generates following output, which lacks of all EA-essential context information:
![image](https://cloud.githubusercontent.com/assets/1596462/6043391/3b0e20f4-ac8c-11e4-8454-8f2dceee0efb.png)

When using strongly typed version instead:

```@Html.DropDownListFor(model => model.CarModelID, null, "--select--", htmlAttributes: new { @class = "form-control" })```

you get following HTML:
![image](https://cloud.githubusercontent.com/assets/1596462/6043398/4458ef0e-ac8c-11e4-80f5-6410df130794.png)

Strongly typed helpers were introduced in version 2.0 of ASP.NET MVC. Since then there is no point of using older versions. You gain a lot - strongly typed lambda expression tied to given model property instead of property name hardcoded as a magic string. Out of the box you have intelli-sense, refactoring tools and compiler support.

**Comment by nunos7** [comment]
Yes you are correct, the page was generated by visual studio and i didn't change anything...

Thanks for your time,
Nuno Santos

**Comment by nunos7** [comment]
Is it possible to fire the validation only on the submit click, instead of on the change event at the checkbox?

**Comment by JaroslawWaliszko** [comment]
Yes, it is explained by the following section of the manual: [How to control frequency of dependent fields validation?](https://github.com/JaroslawWaliszko/ExpressiveAnnotations#how-to-control-frequency-of-dependent-fields-validation)

In your case triggers it should be turned of:
```
@section Scripts {
   ...
    <script>
        ea.settings.dependencyTriggers = undefined;
    </script>
}
```

**Comment by nunos7** [comment]
Thanks :),

Nuno Santos

**Comment by JaroslawWaliszko** [comment]
You're welcome,
Regards

**Comment by JaroslawWaliszko** [comment]
I was interested of what is that special in strong helper `DropDownListFor` when compared to weak one `DropDownList` in terms of rendering sufficient HTML output. As shown on the screen captures few posts above, strong helper includes client side validation rules, while weak one does not (at least in your sample).

I've looked into System.Web.Mvc.dll library and found some implementation differences.

Before I'll go through it, just to put the context in one place, we have following situation:

* controller passes drop down items through `ViewBag`:

    ```
    public ActionResult Create()
    {
        ViewBag.CarModelID = new SelectList(db.CarModels, "CarModelID", "CarModelName");
    }
```
* weak helper generates insufficient HTML:	

    ```
    @Html.DropDownList("CarModelID", null, "-- select --", htmlAttributes...) 
```	
* strong helper generates solid HTML with all required attributes:

    ```
    @Html.DropDownListFor(model => model.CarModelID, null, "-- select --", htmlAttributes...)
```

The thing is strong version accepts `Expression<Func<TModel, TProperty>>` lambda expression as a parameter, based on which correct metadata object is easily constructed using `ModelMetadata.FromLambdaExpression(...)` function. Validator providers can be then collected using this metadata object, and finally client validation rules are properly extracted from them.

In weak version on the other hand, there is no lambda expression given. Since there is no explicitly provided expression from which metadata object can be easily obtained, it is created using another method mainly `ModelMetadata.FromStringExpression(...)`. Its implementation extracted from the ASP.NET MVC framework is shown below:
```
internal static ModelMetadata FromStringExpression(string expression, ViewDataDictionary viewData, ModelMetadataProvider metadataProvider)
{
	...
	ViewDataInfo vdi = viewData.GetViewDataInfo(expression);
	object container = null;
	Type containerType = null;
	Type modelType = null;
	Func<object> modelAccessor = null;
	string propertyName = null;

	if (vdi != null)
	{
		if (vdi.Container != null)
		{
			container = vdi.Container;
			containerType = vdi.Container.GetType();
		}

		modelAccessor = () => vdi.Value;

		if (vdi.PropertyDescriptor != null)
		{
			propertyName = vdi.PropertyDescriptor.Name;
			modelType = vdi.PropertyDescriptor.PropertyType;
		}
		else if (vdi.Value != null)
		{
			// We only need to delay accessing properties (for LINQ to SQL)
			modelType = vdi.Value.GetType();
		}
	}
	else if (viewData.ModelMetadata != null)
	{			
		//  Try getting a property from ModelMetadata if we couldn't find an answer in ViewData
		ModelMetadata propertyMetadata = viewData.ModelMetadata.Properties.Where(p => p.PropertyName == expression).FirstOrDefault();
		if (propertyMetadata != null)
		{
			return propertyMetadata;
		}
	}

	return GetMetadataFromProvider(modelAccessor, modelType ?? typeof(string), propertyName, container, containerType, metadataProvider);
}
```
Unfortunately, while debugging through it, I've noticed that the model metadata object is incorrectly constructed. The code execution enters the first `if` block and returns some weird object, based on which expected client validation rules can not be properly extracted.
				
Without digging too deep, first possibly liable solution would be to swap the orders of if blocks		
```		
    if (viewData.ModelMetadata != null)
    {
	    ...
    }
    else if (vdi != null)
    {
        ...
    }
```		
which gives appropriate HTML output in both of our cases.

To summarize, usage of dynamic `ViewBag` as a container for items implicitly feeding drop down list is really obscure idea and I'd rather avoid that. That's why I strongly recommend usage of the construction I mentioned earlier in this thread, mainly:
```
@Html.DropDownListFor(model => model.CarModelID, Model.Feeds, "-- select --", htmlAttributes...)
```
(where `Feeds` is of type `IEnumerable<SelectListItem>` and initialized as `new SelectList(db.CarModels, "CarModelID", "CarModelName")`)

Not only it omits extraction of data from `ViewBag` (which is not needed at all this time), but in addition uses lambda expression which has many advantages already discussed earlier.
